### DISCUSSION TITLE
Helper function to compare queryKeys

### DISCUSSION THREAD

**Discussion by JonKrone** [TanStack/query#4355] [Q&A]
Hi there,

Does ReactQuery have an officially sanctioned way to use the same internal mechanism the library does to compare queryKeys? I've had a need for this a few times over the year or so I've used RQ - particularly when building my own query matching logic for the `predicate` query filter.

After a little digging, it seems that [this snippet here is what I'm looking for](https://github.com/TanStack/query/blob/main/packages/query-core/src/utils.ts#L292), but is not an available import (because it's not listed in the top-level package.json exports field).

An example usecase: Say you have a multi-tenant app like Stripe where you have one Account that you log in with but also have ownership of many different projects and can view those one at a time. When you switch between those projects, you may want to clear the cache of all the queries related to the previous project but keep the high-level stuff like your Account query and the list of projects. So:

```tsx
const keysToKeep = [
  QueryKeys.currentAccount(), // i.e. ['currentAccount']
  QueryKeys.projects(), // i.e ['projects']
  QueryKeys.project(), // i.e. ['project']
]

const resetQueryCache = () => {
  queryClient.cancelQueries()

  return queryClient.invalidateQueries({
    predicate: ({ queryKey }) => {
      // Here we'd use the queryKey match helper to determine
      // if the query should be kept
    },
  })
}

const handleProjectChange = () => {
  resetQueryCache()
  // .. your other logic
}
```

**Comment by TkDodo** [comment]
I've thought about this as well. for `predicate`, it would be nice if we'd export our `partialMatchKey` function:

https://github.com/TanStack/query/blob/f9ebf9aaae59050331c22f143d835e46fac402cc/packages/query-core/src/utils.ts#L282-L287

I'd accept a PR that does this with proper documentation for the use-case.
